source/git-api.js

```diff
@@ -31,87 +30,126 @@ exports.registerApi = (env) => {
       socket.on('disconnect', () => {
         stopDirectoryWatch(socket);
       });
-      socket.on('watch', (data, callback) => {
+      socket.on('watch', async (data) => {
         stopDirectoryWatch(socket); // clean possibly lingering connections
         socket.watcherPath = path.normalize(data.path);
         socket.join(socket.watcherPath); // join room for this path
 
-        fs.readFile(path.join(socket.watcherPath, '.gitignore'))
-          .then((ignoreContent) => (socket.ignore = ignore().add(ignoreContent.toString())))
-          .catch(() => {})
-          .then(() => {
-            socket.watcher = [];
-            return watchPath(socket, '.', { recursive: isMac || isWindows });
-          })
-          .then(() => {
-            if (!isMac && !isWindows) {
-              // recursive fs.watch only works on mac and windows
-              const promises = [];
-              promises.push(watchPath(socket, path.join('.git', 'HEAD')));
-              promises.push(watchPath(socket, path.join('.git', 'refs', 'heads')));
-              promises.push(watchPath(socket, path.join('.git', 'refs', 'remotes')));
-              promises.push(watchPath(socket, path.join('.git', 'refs', 'tags')));
-              return Promise.all(promises);
-            }
-          })
-          .finally(callback);
+        const watcher = await watchRepo(socket.watcherPath);
+        watcher.on(
+          'workdir',
+          _.throttle((changedPath) => {
+            winston.info(`${changedPath} triggered workdir refresh for ${socket.watcherPath}`);
+            emitWorkingTreeChanged(socket.watcherPath);
+          }, 200)
+        );
+        watcher.on(
+          'git',
+          _.throttle((changedPath) => {
+            winston.info(`${changedPath} triggered git refresh for ${socket.watcherPath}`);
+            emitGitDirectoryChanged(socket.watcherPath);
+          }, 200)
+        );
+        watcher.on('error', (err) => {
+          winston.warn(`Error watching ${socket.watcherPath}: `, JSON.stringify(err));
+        });
+        socket.watcher = watcher;
       });
     });
   }
 
-  const watchPath = (socket, subfolderPath, options) => {
-    const pathToWatch = path.join(socket.watcherPath, subfolderPath);
+  let watcherId = 1;
+  class RepoWatcher extends EventEmitter {
```

Comment:
nit: this file is one of those file that is getting too big so it might be nice to extract this out into a separate file.

Reply:
yes, that's the idea on the nodegit branch but I can only do that when this PR gets merged